From 60e156e3fda8b6b73d1949a9722f5c3d71e96680 Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:19 +0800
Subject: [PATCH 7/9] GSI Android 17 compatibility:
 0007-Add-a-property-to-cap-declared-le-vendor-version.-Fo.patch

Change-Id: I0f23c8d92ea97401563e44e86143d684c78431e7
---
 system/gd/hci/controller_impl.cc | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/system/gd/hci/controller_impl.cc b/system/gd/hci/controller_impl.cc
index 0fdc172..2c5b480 100644
--- a/system/gd/hci/controller_impl.cc
+++ b/system/gd/hci/controller_impl.cc
@@ -684,6 +684,13 @@ struct ControllerImpl::impl {
     }
     vendor_capabilities_.is_supported_ = 0x01;
 
+    int vendor_cap_max = 0xffff;
+    std::string vendor_cap_max_prop = GetSystemProperty("persist.sys.bt.max_vendor_cap").value_or("");
+    if (vendor_cap_max_prop != "") {
+        vendor_cap_max = std::stoi(vendor_cap_max_prop);
+    }
+    if (vendor_cap_max < 55) return;
+
     // v0.55
     BaseVendorCapabilities base_vendor_capabilities = complete_view.GetBaseVendorCapabilities();
     vendor_capabilities_.max_advt_instances_ = base_vendor_capabilities.max_advt_instances_;
@@ -703,6 +710,8 @@ struct ControllerImpl::impl {
       return;
     }
 
+    if (vendor_cap_max < 95) return;
+
     // v0.95
     auto v95 = LeGetVendorCapabilitiesComplete095View::Create(complete_view);
     if (!v95.IsValid()) {
@@ -719,6 +728,8 @@ struct ControllerImpl::impl {
       return;
     }
 
+    if (vendor_cap_max < 96) return;
+
     // v0.96
     auto v96 = LeGetVendorCapabilitiesComplete096View::Create(v95);
     if (!v96.IsValid()) {
@@ -733,6 +744,8 @@ struct ControllerImpl::impl {
       return;
     }
 
+    if (vendor_cap_max < 98) return;
+
     // v0.98
     auto v98 = LeGetVendorCapabilitiesComplete098View::Create(v96);
     if (!v98.IsValid()) {
